common/src/Assets/DecalDefinition.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+ Copyright (C) 2023 Kristian Duske
+
+ This file is part of TrenchBroom.
+
+ TrenchBroom is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ TrenchBroom is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with TrenchBroom. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "DecalDefinition.h"
+
+#include "EL/EvaluationContext.h"
+#include "EL/Expressions.h"
+#include "EL/Types.h"
+#include "EL/Value.h"
+#include "EL/VariableStore.h"
+
+#include <kdl/reflection_impl.h>
+#include <kdl/string_compare.h>
+
+#include <vecmath/scalar.h>
+#include <vecmath/vec_io.h>
+
+namespace TrenchBroom
+{
+namespace Assets
+{
+
+namespace
+{
+std::string textureName(const EL::Value& value)
+{
+  if (value.type() != EL::ValueType::String)
+  {
+    return {};
+  }
+  const std::string& texture = value.stringValue();
+  return texture;
+}
+
+DecalSpecification convertToDecal(const EL::Value& value)
+{
+  switch (value.type())
+  {
+  case EL::ValueType::Map:
+    return DecalSpecification{textureName(value[DecalSpecificationKeys::Texture])};
+  case EL::ValueType::String:
+    return DecalSpecification{textureName(value)};
+  case EL::ValueType::Boolean:
+  case EL::ValueType::Number:
+  case EL::ValueType::Array:
+  case EL::ValueType::Range:
+  case EL::ValueType::Null:
+  case EL::ValueType::Undefined:
+    break;
+  }
+
+  return DecalSpecification{};
+}
+} // namespace
+
+DecalSpecification::DecalSpecification()
+  : textureName{""}
+{
+}
+
+DecalSpecification::DecalSpecification(const std::string& i_textureName)
+  : textureName{i_textureName}
+{
+}
+
+kdl_reflect_impl(DecalSpecification);
+
+DecalDefinition::DecalDefinition()
+  : m_expression{EL::LiteralExpression{EL::Value::Undefined}, 0, 0}
+{
+}
+
+DecalDefinition::DecalDefinition(const size_t line, const size_t column)
+  : m_expression{EL::LiteralExpression{EL::Value::Undefined}, line, column}
+{
+}
+
+DecalDefinition::DecalDefinition(const EL::Expression& expression)
+  : m_expression{expression}
+{
+}
+
+void DecalDefinition::append(const DecalDefinition& other)
```

Comment:
I'm not sure that you need all this flexibility. Do you need TB expressions for the decals? Do you want them to be able to depend on entity properties etc?. And how do these decals behave wrt. inheritance?

Reply:
The extra flexibility isn't needed for HL1 support, my thinking with this was to make it easier to extend with things that other engines might need without needing to do more refactoring, much like how the model definition works. (say for example, an engine might have some extra flags to pass through to the renderer such as transparency mode, scale, offsets, etc. Just a hypothetical example, this isn't needed for HL1 specifically)

All the inheritance stuff I just copied across from the model stuff - no entity in HL1 inherits decal properties, so it's not really defined.

The decal will always depends on (at least) one entity property - in Goldsource, it's the one called `texture`. If EL support in the definition is dropped I would at the very least like to be able to have the property name (optionally) specified, for example: `decal()` would be equivalent to `decal(texture)`, but `decal(something_else)` could also be used to reference that property name instead.

Reply:
I see.